core-parser/src/main/java/org/uniprot/core/parser/fasta/uniparc/UniParcProteomeFastaParser.java

```diff
@@ -0,0 +1,134 @@
+package org.uniprot.core.parser.fasta.uniparc;
+
+import org.uniprot.core.Property;
+import org.uniprot.core.uniparc.UniParcCrossReference;
+import org.uniprot.core.uniparc.UniParcDatabase;
+import org.uniprot.core.uniparc.UniParcEntry;
+import org.uniprot.core.uniprotkb.taxonomy.Organism;
+import org.uniprot.core.util.Utils;
+
+import java.util.*;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static org.uniprot.core.parser.fasta.FastaUtils.parseSequence;
+import static org.uniprot.core.util.Utils.*;
+
+public class UniParcProteomeFastaParser {
+
+    private static final Set<UniParcDatabase> uniProtDatabases = Set.of(
+            UniParcDatabase.SWISSPROT, UniParcDatabase.TREMBL, UniParcDatabase.SWISSPROT_VARSPLIC);
+
+    public static String toFasta(UniParcEntry entry) {
+        String id = entry.getUniParcId().getValue();
+
+        List<UniParcCrossReference> uniProtXrefs = new ArrayList<>();
+        List<UniParcCrossReference> sourceXrefs = new ArrayList<>();
+        boolean hasActive = false;
+        boolean hasSourceActive = false;
+        String proteomeId = null;
+        for(UniParcCrossReference xref: entry.getUniParcCrossReferences()){
+            if(uniProtDatabases.contains(xref.getDatabase())){
+                uniProtXrefs.add(xref);
+                if(proteomeId == null && xref.hasProperties()){
+                    proteomeId = getProteomeId(xref.getProperties().get(0));
+                }
+                if(xref.isActive()){
+                    hasActive = true;
+                }
+            } else {
+                sourceXrefs.add(xref);
+                proteomeId = xref.getProteomeId();
+                if(xref.isActive()){
+                    hasSourceActive = true;
+                }
+            }
+        }
```

Comment:
You are assigning proteomeId twice, one in uniProtDatabases if and then in sourceXrefs else block.
After that you are using it is header, will it be an issue if we get proteomeId from source but has uniRefXref?

To make code clear i would suggest to traverse the crossreferences twice, once to compute uniProtXrefs. Traverse again to get sourceXrefs if uniProtXrefs is empty.